mlos_core/mlos_core/optimizers/observations.py

```diff
@@ -0,0 +1,361 @@
+
+        Parameters
+        ----------
+        index : pd.Index
+            The indices to filter by.
+
+        Returns
+        -------
+        Observation
+            The filtered observation.
+        """
+        return Observations(
+            config=self.config.loc[index],
+            score=self.score.loc[index],
+            context=None if self.context is None else self.context.loc[index],
+            metadata=None if self.metadata is None else self.metadata.loc[index],
+        )
+
+    def append(self, other: Observation) -> None:
+        """
+        Appends the given observation to this observation.
+
+        Parameters
+        ----------
+        other : Observation
+            The observation to append.
+        """
+        config = other.config.to_frame().T
+        score = other.score.to_frame().T
+        context = None if other.context is None else other.context.to_frame().T
+        metadata = None if other.metadata is None else other.metadata.to_frame().T
+        if len(self.config.index) == 0:
+            self.config = config
+            self.score = score
+            self.context = context
+            self.metadata = metadata
+            assert set(self.config.index) == set(
+                self.score.index
+            ), "config and score must have the same index"
+            return
+
+        self.config = pd.concat([self.config, config]).reset_index(drop=True)
+        self.score = pd.concat([self.score, score]).reset_index(drop=True)
+        assert set(self.config.index) == set(
+            self.score.index
+        ), "config and score must have the same index"
+
+        if self.context is not None:
+            assert context is not None, (
+                "context of appending observation must not be null"
+                + " if context of prior observation is not null"
+            )
+            self.context = pd.concat([self.context, context]).reset_index(drop=True)
+            assert self.config.index.equals(
+                self.context.index
+            ), "config and context must have the same index"
+        else:
+            assert context is None, (
+                "context of appending observation must be null"
+                + " if context of prior observation is null"
+            )
+        if self.metadata is not None:
+            assert metadata is not None, (
+                "context of appending observation must not be null"
+                + " if metadata of prior observation is not null"
+            )
+            self.metadata = pd.concat([self.metadata, metadata]).reset_index(drop=True)
+            assert self.config.index.equals(
+                self.metadata.index
+            ), "config and metadata must have the same index"
+        else:
+            assert metadata is None, (
+                "context of appending observation must be null"
+                + " if metadata of prior observation is null"
```

Comment:
```suggestion
                " if metadata of prior observation is null"
```

same comment about the trailing whitespace on the previous line here.